src/sail_lean_backend/pretty_print_lean.ml

```diff
@@ -198,23 +257,35 @@ let rec doc_exp (E_aux (e, (l, annot)) as full_exp) =
   | E_internal_plet _ -> string "sorry" (* TODO replace by actual implementation of internal_plet *)
   | E_app (f, args) ->
       let d_id =
-        if Env.is_extern f env "lean" then string (Env.get_extern f env "lean") else doc_exp (E_aux (E_id f, (l, annot)))
+        if Env.is_extern f env "lean" then string (Env.get_extern f env "lean")
+        else doc_exp ctxt (E_aux (E_id f, (l, annot)))
       in
-      let d_args = List.map doc_exp args in
+      let d_args = List.map (doc_exp ctxt) args in
       nest 2 (parens (flow (break 1) (d_id :: d_args)))
   | E_vector vals -> failwith "vector found"
-  | E_typ (typ, e) -> parens (separate space [doc_exp e; colon; doc_typ typ])
-  | E_tuple es -> parens (separate_map (comma ^^ space) doc_exp es)
+  | E_typ (typ, e) -> parens (separate space [doc_exp ctxt e; colon; doc_typ ctxt typ])
+  | E_tuple es -> parens (separate_map (comma ^^ space) (doc_exp ctxt) es)
   | E_let (LB_aux (LB_val (lpat, lexp), _), e) ->
       let id =
         match pat_is_plain_binder env lpat with
         | Some (Some (Id_aux (Id id, _))) -> id
         | Some None -> "x" (* TODO fresh name or wildcard instead of x *)
         | _ -> failwith "Let pattern not translatable yet."
       in
-      nest 2 (flow (break 1) [string "let"; string id; coloneq; doc_exp lexp]) ^^ hardline ^^ doc_exp e
+      nest 2 (flow (break 1) [string "let"; string id; coloneq; doc_exp ctxt lexp]) ^^ hardline ^^ doc_exp ctxt e
   | _ -> failwith ("Expression " ^ string_of_exp_con full_exp ^ " " ^ string_of_exp full_exp ^ " not translatable yet.")
 
+let doc_binder ctxt i t =
+  let paranthesizer =
+    match t with
+    | Typ_aux (Typ_app (Id_aux (Id "implicit", _), [A_aux (A_nexp (Nexp_aux (Nexp_var ki, _)), _)]), _) ->
+        implicit_parens
+    | _ -> parens
+  in
+  (* Overwrite the id if foo *)
```

Comment:
This comment is unfinished?

Reply:
Thanks, if only I remembered what I wanted to put there :grimacing: